phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexChecker.java

```diff
@@ -0,0 +1,371 @@
+            buildIndexScan.setTimeRange(ts, maxTimestamp);
+            buildIndexScan.setRaw(true);
+            ResultScanner resultScanner = dataHTable.getScanner(buildIndexScan);
+            resultScanner.next();
+            // Close the current scanner as the newly build row will not be visible to it
+            scanner.close();
+            // Open a new scanner starting from the current row
+            indexScan.withStartRow(indexRowKey, true);
+            scanner = region.getScanner(indexScan);
+            // Scan the newly build index rows
+            scanner.next(row);
+            if (row.isEmpty()) {
+                return;
+            }
+            // Check if the corresponding data table row exist
+            if (Bytes.compareTo(row.get(0).getRowArray(), row.get(0).getRowOffset(), row.get(0).getRowLength(),
+                    indexRowKey, 0, indexRowKey.length) == 0) {
+                if (!verifyRowAndRemoveVerifiedCell(row)) {
+                    // The corresponding row does not exist in the data table.
+                    // Need to delete the row from index if it is old enough
+                    deleteRowIfAgedEnough(indexRowKey, ts);
+                    row.clear();
+                }
+                return;
+            }
+            // This means the current index row is deleted by the rebuild process and we got the next row.
+            // If it is verified then we are good to go. If not, then we need to repair the new row
+            if (!verifyRowAndRemoveVerifiedCell(row)) {
+                // Rewind the scanner and let the row be scanned again so that it can be repaired
+                scanner.close();
+                scanner = region.getScanner(indexScan);
+                row.clear();
+            }
+        }
+
+        private boolean isVerifiedCell(Cell cell) {
+            return Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(), cell.getFamilyLength(),
+                    emptyCF, 0, emptyCF.length) == 0 &&
+                    Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(), cell.getQualifierLength(),
+                            emptyCQ, 0, emptyCQ.length) == 0;
+        }
+
+        private boolean verifyRow(byte[] rowKey) throws IOException {
+            Get get = new Get(rowKey);
+            get.setTimeRange(minTimestamp, maxTimestamp);
+            get.addColumn(emptyCF, emptyCQ);
+            Result result = region.get(get);
+            if (result.isEmpty()) {
+                LOG.debug("The empty column does not exist in a row in " + region.getRegionInfo().getTable().getNameAsString());
+                return true;
+            }
+            LOG.debug("Scan " + scan + " did not return the empty column for " + region.getRegionInfo().getTable().getNameAsString());
+            if (Bytes.compareTo(result.getValue(emptyCF, emptyCQ), 0, FALSE_BYTES.length,
+                    FALSE_BYTES, 0, FALSE_BYTES.length) == 0) {
```

Comment:
Please add comments in code as to why FALSE_BYTES is used here instead of TRUE_BYTES or something else.

Reply:
Will do that

Reply:
Done